### PR TITLE
chore(gux-datepicker): region specific date formatting

### PR DESCRIPTION
**Pull request by katie-bobbe-genesys**
Date formatting should get now supplement region data from the browser.
This behavior is "feature toggled" using a cookie, called spark-enable-regional-dates. To test out the functionality, set document.cookie = 'spark-enable-regional-dates' in your console. I opened a separate browser and set the region to "en-GB"  in my system preferences. To check the region in your console, just type navigator.language. 

The date formats should now be formatted according to locale.

### CODE REVIEW

File: packages/genesys-spark-components/src/components/stable/gux-calendar/gux-calendar.service.ts [hunk 1]
```diff
@@ -1,4 +1,6 @@
 import { capitalizeFirstLetter } from '@utils/string/capitalize-first-letter';
+import * as sparkIntl from '../../../genesys-spark-utils/intl';
+import { readRegionalDatesCookie } from '../../../i18n/check-regional-dates-cookie';
```
Comment (daragh-king-genesys) [inline packages/genesys-spark-components/src/components/stable/gux-calendar/gux-calendar.service.ts:3]: Should we comment these line with a link to https://inindca.atlassian.net/browse/COMUI-2598 to make that issue more visible?

File: packages/genesys-spark-components/src/components/stable/gux-datepicker/gux-datepicker.service.ts [hunk 2]
```diff
@@ -187,3 +188,37 @@ export function getIntervalRange(
 
   return { selectionStart, selectionEnd };
 }
+
+export function getFormat(component: HTMLElement): string {
```
Comment (daragh-king-genesys) [inline packages/genesys-spark-components/src/components/stable/gux-datepicker/gux-datepicker.service.ts:192]: I think I would move this logic into a function in `sparkIntl` that takes a `locale` not a component. I think that would make it easier to test.
Then this function would look something like 
```
export function getFormat(component: HTMLElement): string {
  const locale = sparkIntl.determineDisplayLocale(component);

  return sparkIntl.determineDisplayLocaleDateFormat(locale); // FYI determineDisplayLocaleDateFormat is a terrible name :)
}
```

File: packages/genesys-spark-components/src/components/stable/gux-datepicker/gux-datepicker.service.ts [hunk 3]
```diff
@@ -187,3 +188,37 @@ export function getIntervalRange(
...
+  const fixedDate = new Date('December 31, 2000 15:00:00 UTC+00:00');
```
Comment (daragh-king-genesys) [inline packages/genesys-spark-components/src/components/stable/gux-datepicker/gux-datepicker.service.ts:193]: There may be an issue with this date and timezones. ie this moment in time is Jan 1st in australia

File: packages/genesys-spark-components/src/components/stable/gux-datepicker/gux-datepicker.service.ts [hunk 4]
```diff
@@ -187,3 +188,37 @@ export function getIntervalRange(
...
+    .dateTimeFormat(sparkIntl.determineDisplayLocale(component), {
+      year: 'numeric'
+    })
+    .format(fixedDate)
+    .match(/(\d+)/)[0];
+
+  return localizedFixedDate
+    .replace(fixedDay, 'dd')
+    .replace(fixedMonth, 'mm')
+    .replace(fixedYear, 'yyyy');
```
Comment (daragh-king-genesys) [inline packages/genesys-spark-components/src/components/stable/gux-datepicker/gux-datepicker.service.ts:223]: ave a look at this
```
const date = new Date('July 5, 2000 15:00:00 UTC+00:00');
const options = {
 year: 'numeric',
  month: 'numeric',
  day: 'numeric'
};
const dateTimeFormat = new Intl.DateTimeFormat('ru', options);

const parts = dateTimeFormat.formatToParts(date);

console.log(parts);
```

File: packages/genesys-spark-components/src/components/stable/gux-datepicker/gux-datepicker.tsx [hunk 5]
```diff
@@ -645,7 +678,15 @@ export class GuxDatepicker {
   async componentWillLoad() {
     trackComponent(this.root, { variant: this.mode });
     this.i18n = await buildI18nForComponent(this.root, translationResources);
-
+    if (readRegionalDatesCookie()) {
+      if (!this.format) {
+        this.format = getFormat(this.root);
+      } else {
+        this.formatOverride = true;
```
Comment (daragh-king-genesys) [inline packages/genesys-spark-components/src/components/stable/gux-datepicker/gux-datepicker.tsx:685]: Is this overly complicated?
```
if (readRegionalDatesCookie()) {
  this.format = this.format || getFormat(this.root);
} else {
  this.format = this.format || 'mm/dd/yyyy';
}
```
and then no change to `updateDate`.

File: packages/genesys-spark-components/src/components/stable/gux-datepicker/gux-datepicker.tsx [hunk 6]
```diff
@@ -542,23 +545,53 @@ export class GuxDatepicker {
...
+      } else {
+        const dateValue = fromIsoDate(this.value);
+        this.formattedValue = sparkIntl
+          .dateTimeFormat(sparkIntl.determineDisplayLocale(this.root), {
+            year: 'numeric',
+            month: 'numeric',
+            day: 'numeric'
+          })
+          .format(new Date(dateValue));
+      }
```
Comment (gavin-everett-genesys) [inline packages/genesys-spark-components/src/components/stable/gux-datepicker/gux-datepicker.tsx:574]: Would this be more readable as, I have not tested it fully so there may be errors,

```
if(regionalDatesCookie() && !this.formatOvveride {
   const locale = sparkIntl.determineDisplayLocale(this.root);
  const options = {
  year: 'numeric',
 month: 'numeric',
 day: 'numeric',
}

const sparkIntlFormatter = sparkIntl.dateTimeFormat(locale, options);

if(this.mode === CalendarModes.Range) {
 const [from, to] = fromIsoDateRange(this.value);
const fromDate = new Date(from);
const toDate = new Date(to);

this.formattedValue = sparkIntlFormatter.format(fromDate);
this.toFormattedValue = sparkIntlFormatter.format(toDate);
} else {
 const dateValue = fromIsoDate(this.value);
const formatDate = sparkIntlFormatter.format(new Date(dateValue));
this.formattedValue = formattedDate;

}
}
```

File: packages/genesys-spark/src/utils/intl.ts [hunk 7]
```diff
@@ -86,3 +86,33 @@ function browserHasRegionData(localeString: string): boolean {
...
+        case 'month':
+          return `mm`;
+        case 'year':
+          return `yyyy`;
+        default:
+          return value;
+      }
+    })
+    .join('');
+  return dateString.replace(/\s/g, '');
```
Comment (katie-bobbe-genesys) [inline packages/genesys-spark/src/utils/intl.ts:117]: date strings with spaces do not work with the current keyboard navigation. I'm wondering if removing the spaces is too much interference with the regional date format?

**Comment by katie-bobbe-genesys** [comment]
Todo: The `ar` date format does not work as expected

File: packages/genesys-spark/src/utils/intl.ts [hunk 8]
```diff
@@ -86,3 +86,34 @@ function browserHasRegionData(localeString: string): boolean {
...
+          return `mm`;
+        case 'year':
+          return `yyyy`;
+        default:
+          return value;
+      }
+    })
+    .join('');
+  // review locales with two character date delimiters https://inindca.atlassian.net/browse/COMUI-2679
+  return dateString.replace(/\s/g, '').replace(/‏/g, '');
```
Comment (katie-bobbe-genesys) [inline packages/genesys-spark/src/utils/intl.ts:118]: this is the character used in the delimiter in Arabic https://unicode-explorer.com/c/200F

**Review by daragh-king-genesys — APPROVED** [review]

**Pull request merged**